lib/src/merge.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2023 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use itertools::Itertools;
+
+/// Attempt to resolve trivial conflicts between the inputs. There must be
+/// exactly one more adds than removes.
+pub fn trivial_merge<T>(removes: &[T], adds: &[T]) -> Option<T>
+where
+    T: Eq + Hash + Clone,
+{
+    assert_eq!(
+        adds.len(),
+        removes.len() + 1,
+        "trivial_merge() requires exactly one more adds than removes"
+    );
+
+    // Check if all sides made the same change.
+    // This matches what Git and Mercurial do (in the 3-way case at least), but not
+    // what Darcs and Pijul do. It means that repeated 3-way merging of multiple
+    // trees may give different results depending on the order of merging.
+    if adds.iter().all_equal() && removes.iter().all_equal() {
```

Comment:
> I think the first one should at least be a conflict.

Let me know if you can remember the reason. To me, it seems that if we resolve `[1], [2,2]` to 2, resolving `[0, 1], [2, 2, 2]` to `[0], [2,2]` and then to 2 is pretty natural.

Also, it might be nice to experiment with having more and less lenient merge algorithms so that we can compare. That would take creating some UI, though.

Reply:
> Let me know if you can remember the reason. To me, it seems that if we resolve `[1], [2,2]` to 2, resolving `[0, 1], [2, 2, 2]` to `[0], [2,2]` and then to 2 is pretty natural.

In the case of `[1], [2,2]`, it represents two identical changes. Since `[0,1], [2,2,2]` represents two different changes, it seems riskier to say that they don't conflict. Either way, wanted to be as conservative as practically possible, and complex cases like this seemed more okay to error out on.

> Also, it might be nice to experiment with having more and less lenient merge algorithms so that we can compare. That would take creating some UI, though.

As it happens, @chooglen and I talked about that last week. We continued the discussion we had on Discord about `X+(X-N)` conflicts and the more complex case you mentioned where the order of merging would produce different  results. I think it would be reasonable to have a strict merge strategy that doesn't automatically resolve any of these weird cases (like `X+(X-N)`), leaving it as a conflict in the commit. Then if the user merge that commit with another change that with a `N-X` change, it would become resolved. I believe it would solve the case you brought up. Then we could have a more lenient merge strategy that we use when we materialize conflicts, so when checking out a commit with such a conflict, we would automatically resolve it in the working copy. If we did that, then running `jj diff` right after `jj checkout` would show that a conflict was resolved in this case. I think Glen and I agreed that it would make some sense to do that but that it wouldn't be worth it because users would usually don't care and it would be just be annoying to have to tell jj to do the automatic resolution.

Reply:
(Note: I meant to proofread this before posting, but I'm about to go to sleep and it's relevant for an ongoing Discord discussion, so I'll post it now)

The simplest UI would be to have a `jj resolve --auto` command that resolves a conflict using a lax automerge strategy.

Here's an example of a modification to the approach you described: `jj` could remember an "automerge strategy" for each file path as part of a change. As in your example, the entire conflict for a path would also be stored unless the strictest strategy doesn't resolve it. The strictest strategy could require all the removes and all the adds to be identical.

However, git export and the working copy would use a (more lenient) automerge strategy to resolve the conflict further. If that strategy turns out to be too lax, there would be a command for the user to change the automerge strategy for that file to something more strict. This would be stored in the jj change, and would also rewrite the git commit id and the way that commit would be checked out in a worktree. `jj diff` would show diffs using the selected automerge strategies as opposed to showing every automerge as a conflict resolution. It could potentially show if a paths automerge strategy was changed.